src/binxml/assemble.rs

```diff
@@ -282,119 +265,120 @@ pub fn create_record_model<'a>(
-                    Cow::Owned(BinXMLDeserializedTokens::Value(Cow::Owned(value))),
-                    stack,
-                );
-            }
-        } else {
-            _expand_templates(token, stack);
-        }
+    let value = template
+        .substitution_array
+        .get_mut(substitution_descriptor.substitution_index as usize);
+
+
+    if let Some(value) = value {
+        let value = mem::replace(value,
+                                 BinXMLDeserializedTokens::Value(BinXmlValue::NullType));
+        _expand_templates(
+            Cow::Owned(value),
+            chunk,
+            stack,
+        );
+    } else {
+        _expand_templates(
+            Cow::Owned(BinXMLDeserializedTokens::Value(BinXmlValue::NullType)),
+            chunk,
+            stack,
+        );
     }
 }
 
-fn expand_borrowed_template<'a>(
-    template: &'a BinXmlTemplate<'a>,
+fn expand_template<'a>(
+    mut template: BinXmlTemplate<'a>,
+    chunk: &'a EvtxChunk<'a>,
     stack: &mut Vec<Cow<'a, BinXMLDeserializedTokens<'a>>>,
 ) {
-    // Here we can always use refs, since even if the definition is owned by the template,
-    // we do not own it.
-    for token in template.definition.as_ref().tokens.iter() {
-        if let BinXMLDeserializedTokens::Substitution(ref substitution_descriptor) = token {
-            if substitution_descriptor.ignore {
-                continue;
-            } else {
-                let value = &template
-                    .substitution_array
-                    .get(substitution_descriptor.substitution_index as usize)
-                    .unwrap_or(BinXmlValue::NullType.borrow());
-
-                _expand_templates(
-                    Cow::Owned(BinXMLDeserializedTokens::Value(Cow::Borrowed(value))),
-                    stack,
-                );
+    let template_def = if let Some(template_def) = chunk.template_table.get_template(template.template_def_offset) {
+        template_def
+    } else {
+        // There can be a template which was not found in the header, if the file was not closed correctly.
+        // In that case, we will try to read it directly from the chunk.
+
+        debug!("Template in offset {} was not found in cache", template.template_def_offset);
+        let mut cursor = Cursor::new(chunk.data);
+        let _ = cursor.seek(SeekFrom::Start(u64::from(template.template_def_offset)));
+        let template_def = read_template_definition(&mut cursor, Some(chunk), chunk.settings.get_ansi_codec());
+
+        if let Ok((template_def, _)) = template_def {
+            for token in template_def.tokens {
+                if let BinXMLDeserializedTokens::Substitution(ref substitution_descriptor) = token {
+                    expand_token_substitution(&mut template, substitution_descriptor, chunk, stack)
+                } else {
+                    _expand_templates(Cow::Owned(token), chunk, stack);
+                }
             }
+        }
+        return;
```

Comment:
we kind of swallow the errors here. Should this function be able to propogate failures, or is it OK to label this entire section as `carving` behavior - where we cannot gurantee anything, and which is this case we should probably warn the user that this record might be incomplete.